### PR TITLE
Improve AI's Skill Swap handling in double battles

### LINKED ISSUE #5222: AI does not recognize benefits from Contrary
### Description

> The AI did not recognize its partner benefited from Contrary and preferred to use other moves. When it only had Skill Swap, it gave Contrary to the player. 

# Flags Used
`AI_FLAG_CHECK_BAD_MOVE | AI_FLAG_TRY_TO_FAINT | AI_FLAG_CHECK_VIABILITY | AI_FLAG_HELP_PARTNER`

# AI Team
```
Spinda
Ability: Contrary
- Skill Swap
- Encore
- Fake Tears
- Swagger

Arcanine
Ability: Intimidate
- Extremespeed
- Morning Sun
- Close Combat
- Overheat
```

> The goal is to have Spinda use Skill Swap first turn on Arcanine, giving it Contrary for [Overheat / Close Combat] and triggering Intimidate on the player again. Instead, Spinda [tends] to [use] Swagger [on] the player while Arcanine uses Close Combat.

### Version

master (default, unreleased bugfixes)

### Upcoming/master Version

_No response_

### Discord contact info

Jinnora

### PR DESCRIPTION
**Pull request by Pawkkie**
## Description
Improves how the AI handles Skill Swap alongside both Intimidate and Contrary, as described in #5222. Includes new helper function to identify moves that lower a mon's own stats. 

Minor refactor to the EFFECT_SKILL_SWAP switch case to allow for accumulation of multiple score increases. This is for the specific purpose of having the situations like the test case (Arcanine with Overheart +3 from giving Contrary and Intimidate for +2 for second activation) be a big enough score increase to be preferred over just swapping the mon's ability for a highly rated ability on the player's side (such as Shadow Tag) which would be prioritized otherwise.

## Issue(s) that this PR fixes
Fixes #5222 

## **Discord contact info**
@Pawkkie

### CODE REVIEW

File: src/battle_ai_util.c [hunk 1]
```diff
@@ -2285,6 +2299,34 @@ bool32 IsStatLoweringEffect(u32 effect)
     }
 }
 
+bool32 IsOwnStatLoweringEffect(u32 effect)
```
Comment (AlexOn1ine) [inline src/battle_ai_util.c:2302]:
```suggestion
bool32 IsSelfStatLoweringEffect(u32 effect)
```

File: src/battle_ai_util.c [hunk 2]
```diff
@@ -2080,6 +2080,20 @@ bool32 HasAnyKnownMove(u32 battlerId)
...
 
+bool32 HasMoveThatLowersOwnStats(u32 battlerId)
+{
+    s32 i;
+    u16 *moves = GetMovesArray(battlerId);
+    for (i = 0; i < MAX_MON_MOVES; i++)
+    {
+        if (moves[i] != MOVE_NONE && moves[i] != MOVE_UNAVAILABLE
+            && gMovesInfo[moves[i]].additionalEffects[i].self
+            && IsOwnStatLoweringEffect(gMovesInfo[moves[i]].additionalEffects[i].moveEffect))
```
Comment (AlexOn1ine) [inline src/battle_ai_util.c:2091]: The `additionalEffects` part isn't correct. it needs it's own loop to loop over the effects a move can have

Comment (Pawkkie) [reply src/battle_ai_util.c:2091]: Good catch ty!

File: src/battle_ai_util.c [hunk 3]
```diff
@@ -2080,6 +2080,24 @@ bool32 HasAnyKnownMove(u32 battlerId)
...
+    s32 i, j;
+    u16 *moves = GetMovesArray(battlerId);
+    for (i = 0; i < MAX_MON_MOVES; i++)
+    {
+        if (moves[i] != MOVE_NONE && moves[i] != MOVE_UNAVAILABLE)
+        {
+            for (j = 0; j < gMovesInfo[moves[i]].numAdditionalEffects; j++)
+            {
+                if (IsSelfStatLoweringEffect(gMovesInfo[moves[i]].additionalEffects[j].moveEffect) && gMovesInfo[moves[i]].additionalEffects[j].self)
+                    return TRUE;
```
[the remark this answers is not in the document]
Comment (Pawkkie) [reply src/battle_ai_util.c]: Sure, pushed :)

File: src/battle_ai_util.c [hunk 4]
```diff
@@ -2080,6 +2080,25 @@ bool32 HasAnyKnownMove(u32 battlerId)
     return FALSE;
 }
 
+bool32 HasMoveThatLowersOwnStats(u32 battlerId)
+{
+    s32 i, j;
+    u16 *moves = GetMovesArray(battlerId), aiMove;
+    for (i = 0; i < MAX_MON_MOVES; i++)
```
Comment (AlexOn1ine) [inline src/battle_ai_util.c:2087]:
```suggestion
    u32 aiMove;
    u16 *moves = GetMovesArray(battlerId);
    for (i = 0; i < MAX_MON_MOVES; i++)
```

**Review by AlexOn1ine — APPROVED** [review]

**Pull request merged**